### PR TITLE
Radix Sort

### PR DESCRIPTION
**Pull request by yeoshuheng**
- unit test added
- readme added
- cleanup code to meet checkstyle requirements

### CODE REVIEW

File: src/main/java/algorithms/sorting/radixSort/README.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
+# Radix Sort
+
+Radix Sort is a non-comparison based, stable sorting algorithm. 
+We first treat each element as a string with *w* digits (Padding elements that have
```
Comment (4ndrelim) [inline src/main/java/algorithms/sorting/radixSort/README.md:4]: what is 'w' here?

File: src/main/java/algorithms/sorting/radixSort/README.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
+# Radix Sort
+
+Radix Sort is a non-comparison based, stable sorting algorithm. 
+We first treat each element as a string with *w* digits (Padding elements that have 
+less than *w* digits).
+
+From the least-significant digit to the most-significant digit, we constantly
```
Comment (4ndrelim) [inline src/main/java/algorithms/sorting/radixSort/README.md:7]: this sounds very specific. did you base it off somewhere? and depending on implementation, not necessarily 0-9 only.

Perhaps can talk about splitting into segments, and performing counting sort individually on each segment from left to right. Key here is to have the counting sort sub-routine being stable (can explain why its absolutely necessary for it to be stable too!)

File: src/main/java/algorithms/sorting/radixSort/README.md [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+## Complexity Analysis
+**Time**:
+Let *k* be the base of the number system being used. For integers in java, it is base 10 
+(we have digits 0 to 9), in the case of base 2 (binary), we only have 2 digits (1, 0),
+base 3, (0, 1, 2). Hence, we can see that the base of the number system determines the number of 
+different queues we need to iterate through to rebuild our original array, additionally, we also 
+need to iterate through all *w* positions (digits), this results in the following complexities:
+
+- Worst case: O(w * (n + k))
+- Average case: O(w * (n + k))
```
Comment (4ndrelim) [inline src/main/java/algorithms/sorting/radixSort/README.md:27]: we can discuss a more general version of the formula tmr!

File: src/main/java/algorithms/sorting/radixSort/RadixSort.java [hunk 4]
```diff
@@ -1,53 +1,69 @@
...
-        sorted = arr;
-    }
+  /**
+   * Radix sorts a given input array and updates the output array in-place.
+   *
+   * @param arr original input array.
+   * @param sorted output array.
+   */
+  private static void radixSort(int[] arr, int[] sorted) {
+    // sort the N numbers by segments, starting from left-most segment
```
Comment (4ndrelim) [inline src/main/java/algorithms/sorting/radixSort/RadixSort.java:31]: can also briefly mention the intuition for this in the readme file

File: src/main/java/algorithms/sorting/radixSort/RadixSort.java [hunk 5]
```diff
@@ -1,53 +1,69 @@
...
+      // place each number in its correct sorted position up until the given segment
+      for (int k = arr.length - 1; k >= 0; k--) {
+        int curr = arr[k];
+        int id = getSegmentMasked(curr, i);
+        sorted[freqMap[id] - 1] = curr;
+        freqMap[id]--;
+      }
+      // we do a swap so that our results above for this segment is
+      // saved and passed as input to the next segment
+      int[] tmp = arr;
```
Comment (4ndrelim) [inline src/main/java/algorithms/sorting/radixSort/RadixSort.java:52]: we should mention this is a implementation trick. allowing O(n) space since we are re-using the arrays, one to save our data and another to iteratively work towards the result using the next segment. Swapping back and forth the two as reuse

**Review by 4ndrelim — CHANGES_REQUESTED** [review]

File: src/main/java/algorithms/sorting/radixSort/RadixSort.java (hunk 5 above)
Comment (yeoshuheng) [reply src/main/java/algorithms/sorting/radixSort/RadixSort.java:52]: Added more details

File: src/main/java/algorithms/sorting/radixSort/README.md [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+## Notes
+- Radix sort's time complexity is dependent on the maximum number of digits in each element,
+hence it is ideal to use it on integers with a large range and with little digits.
+- This could mean that Radix Sort might end up performing worst on small sets of data 
+if any one given element has a in-proportionate amount of digits.
+- It is interesting to note that counting sort is used as a sub-routine within the 
+Radix Sort process.
+
+### Common Misconception
+- While not immediately obvious, we can see that radix sort is a stable sorting algorithm as
```
Comment (4ndrelim) [inline src/main/java/algorithms/sorting/radixSort/README.md:60]: conceptual, perhaps can place at the front

File: src/main/java/algorithms/sorting/radixSort/RadixSort.java (hunk 5 above)
Comment (4ndrelim) [reply src/main/java/algorithms/sorting/radixSort/RadixSort.java:56]: yup, in general, one can implement a simpler
O((b/r) * (2^r+n)) space complexity
where we have b/r passes and each pass of counting sort takes (k+n) space where k here is 2^r (range) (assuming we initialize a new array at each pass)

File: src/main/java/algorithms/sorting/radixSort/README.md [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+![Radix Sort](https://miro.medium.com/v2/resize:fit:661/1*xFnpQ4UNK0TvyxiL8r1svg.png)
+
+We place each element into a queue based on the number of possible segments that could be generated.
+Suppose the values of our segments are in base-10, (limited to a value within range *[0, 9]*), 
+we get 10 queues.
+
+*Source: Level Up Coding* 
+
+## Complexity Analysis
```
Comment (4ndrelim) [inline src/main/java/algorithms/sorting/radixSort/README.md:40]: i can help you a little here: 
in general, time complexity is

O((b/r)(n+2^r))
where b-bit word is broken into r-bit pieces and we apply counting sort b/r number of times.
the time complexity of counting sort is O(k+n) where n is the number of elements and k is the range (from 1 to k). here, since each piece is r-bit, range is 2^r.

**Review by 4ndrelim — CHANGES_REQUESTED** [review]
thanks shu heng!

File: src/main/java/algorithms/sorting/radixSort/README.md [hunk 8]
```diff
@@ -0,0 +1,61 @@
+# Radix Sort
+
+Radix Sort is a non-comparison based, stable sorting algorithm. 
+
+Radix Sort continuously sorts based on the least-significant segment of a element
```
[the remark this answers is not in the document]
Comment (yeoshuheng) [reply src/main/java/algorithms/sorting/radixSort/README.md:5]: added

File: src/main/java/algorithms/sorting/radixSort/README.md [hunk 9]
```diff
@@ -0,0 +1,61 @@
+# Radix Sort
```
[the remark this answers is not in the document]
Comment (yeoshuheng) [reply src/main/java/algorithms/sorting/radixSort/README.md:1]: replicated heap readme structure

File: src/main/java/algorithms/sorting/radixSort/README.md (hunk 6 above)
Comment (yeoshuheng) [reply src/main/java/algorithms/sorting/radixSort/README.md:60]: added to front

File: src/main/java/algorithms/sorting/radixSort/RadixSort.java (hunk 4 above)
Comment (yeoshuheng) [reply src/main/java/algorithms/sorting/radixSort/RadixSort.java:32]: if you are referring to the concept of segments and how we segments play a part in radix sort, i think this has already been addressed in the readme in the first few paragraphs

**Comment by yeoshuheng** [comment]
have made changes as per comments

**Review by 4ndrelim — APPROVED** [review]
LGTM. Add ine one more test case focusing on duplicate numbers. feel free to add in more edge cases.
